chain/client/src/sync_jobs_actor.rs

```diff
@@ -149,11 +153,19 @@ impl actix::Handler<WithSpanContext<StateSplitRequest>> for SyncJobsActor {
     fn handle(
         &mut self,
         msg: WithSpanContext<StateSplitRequest>,
-        _: &mut Self::Context,
+        context: &mut Self::Context,
     ) -> Self::Result {
-        let (_span, msg) = handler_debug_span!(target: "client", msg);
-        tracing::debug!(target: "client", ?msg);
-        let response = Chain::build_state_for_split_shards(msg);
-        self.client_addr.do_send(response.with_span_context());
+        let (_span, mut state_split_request) = handler_debug_span!(target: "client", msg);
+        if Chain::retry_build_state_for_split_shards(&state_split_request) {
```

Comment:
The snapshot could theoretically disappear, or be locked between this line and where we try to build states. I think it would be safer to return the locked snapshot and pass it on to build_state_for_split_shards. Implementation may vary, you may be able to just get the lock and pass it or you may need to move this logic to the build_state_for_split_shards.

Reply:
This is a problem in the state sync world as well. `StateSnapshot` struct in tries contains reference to store and flat storage manager and is under a lock. We access this lock to read from or write to state_snapshot.

The stores are cloned and returned to caller and lock is release. However `store` and `flat_storage_manager` are only a reference to the underlying database layer. It's still possible for some thread to delete the snapshot while another thread is holding a cloned reference of the stores.

We see this in trying to access snapshot from ShardTries in `get_trie_with_block_hash_for_shard_from_snapshot` function. Our pattern here is theoretically no different.

Reply:
@nikurt, qq, have we thought about this problem for accessing snapshots?